analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -0,0 +1,433 @@
+﻿using System;
+using System.Text;
+using System.Threading.Tasks;
+
+public class Program
+{
+    private object field = null;
+    private static object staticField = null;
+
+    public void NotCompliantCases(object o, Exception e)
+    {
+        o.ToString(); // FIXME non-compliant {{Refactor this method to add validation of parameter 'o' before using it.}}
+//      ~~~~~~~~~~
+        Bar(o); // Compliant, we care about dereference only
+
+        throw e; // FIXME non-compliant - attempting to throw null as an exception will result in a NullReferenceException
+//      ~~~~~~~
+    }
+
+    public void Bar(object o) { }
+
+    protected void NotCompliantCases_Protected(object o)
+    {
+        o.ToString(); // FIXME non-compliant
+    }
+
+    private void CompliantCases_Private(object o)
+    {
+        o.ToString(); // Compliant, not public
+    }
+
+    protected internal void NotCompliantCases_ProtectedInternal(object o)
+    {
+        o.ToString(); // FIXME non-compliant
+    }
+
+    internal void CompliantCases_Internal(object o)
+    {
+        o.ToString(); // Compliant, not public
+    }
+
+    public void CompliantCases(bool b, object o1, object o2, object o3, object o4, Exception e)
+    {
+        if (o1 != null)
+        {
+            o1.ToString(); // Compliant, we did the check
+        }
+
+        o2 = o2 ?? new object();
+        o2.ToString(); // Compliant, we coalesce
+
+        if (o3 == null)
+        {
+            throw new Exception();
+        }
+
+        o3.ToString(); // Compliant, we did the check
+
+        if (e != null)
+        {
+            throw e; // Compliant
+        }
+
+        o4?.ToString(); // Compliant, conditional operator
+
+        b.ToString(); // Compliant, bool cannot be null
+
+        object v = null;
+        v.ToString(); // Compliant, we don't care about local variables
+
+        field.ToString(); // Compliant
+
+        Program.staticField.ToString(); // Compliant
+    }
+
+    public void MoreCompliantCases(string s1, string s2)
+    {
+        if (string.IsNullOrEmpty(s1))
+        {
+            s1.ToString(); // FIXME non-compliant, could be null
+        }
+        else
+        {
+            s1.ToString(); // Compliant
+        }
+
+        if (string.IsNullOrWhiteSpace(s2))
+        {
+            s2.ToString(); // FIXME non-compliant, could be null
+        }
+        else
+        {
+            s2.ToString(); // Compliant
+        }
+    }
+
+    public void Transitivity(object o1, object o2)
+    {
+        if (o1 == null && o1 == o2)
+        {
+            o2.ToString(); // FIXME non-compliant
```

Comment:
This was not in the old file.
```suggestion
            o2.ToString(); // FN
```

Reply:
Removed the whole method, as all cases inside are covered by S2259.